src/source/folder.js

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2025 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+import { Response } from '@adobe/fetch';
+import { HelixStorage } from '@adobe/helix-shared-storage';
+import { createErrorResponse } from '../contentbus/utils.js';
+import { putSourceFile } from './put.js';
+import { getS3Path, CONTENT_TYPES } from './utils.js';
+
+// A directory is marked by a marker file.
+// This is to allow directories to show up in file listings without
+// having to do a deep S3 listing.
+const FOLDER_MARKER = '_dir';
+const FOLDER_CONTENT_TYPE = 'application/folder';
+
+/**
+ * Convert the directory listing from S3 format to the format expected by the client.
+ *
+ * @param {import('@aws-sdk/client-s3').ListObjectsV2Output} list
+ *   The directory listing as returned by S3.
+ * @returns {Array} The directory listing as returned to the client
+ */
+function transformList(list) {
+  return list.map((item) => {
+    if (item.key.endsWith('/')) {
+      // it's a subfolder
+      return {
+        name: item.path,
+        'content-type': FOLDER_CONTENT_TYPE,
+      };
+    }
+
+    const { lastModified, path } = item;
+    const sp = path.split('.');
+
+    if (sp.length < 2) {
+      // files without extensions are currently not supported
+      return null;
+    }
```

Comment:
could that `.` also occur on a parent directory? and I'd use `splitExtension` (export it first) from `RequestInfo` which would make more of your decisions self-explanatory.